src/pal/src/arch/arm/callsignalhandlerwrapper.S

```diff
@@ -18,13 +18,30 @@ C_FUNC(SignalHandlerWorkerReturnOffset\Alignment):
 // address set to SignalHandlerWorkerReturn during SIGSEGV handling.
 // It enables the unwinder to unwind stack from the handling code to the actual failure site.
 NESTED_ENTRY CallSignalHandlerWrapper\Alignment, _TEXT, NoHandler
-    alloc_stack     (8 + \Alignment) // red zone + alignment
-    PROLOG_PUSH  "{r7, lr}"
+#ifndef __linux__
+__StackAllocationSize = (8 + 4 + \Alignment) // red zone + alignment
+    alloc_stack __StackAllocationSize
+    PROLOG_PUSH "{r7, r11, lr}"
     bl      EXTERNAL_C_FUNC(signal_handler_worker)
 LOCAL_LABEL(SignalHandlerWorkerReturn\Alignment):
-    EPILOG_POP "{r7, lr}"
-    free_stack (8 + \Alignment)    
+    EPILOG_POP "{r7, r11, lr}"
+    // We don't use free_stack here because it lead to compiler generates
+    // ajustment for vsp (vsp = vsp - __StackAllocationSize) before pop
+    // in the unwind info although registers are saved on the top of the stack
+    add sp, sp, __StackAllocationSize
     bx      lr
+#else
+    // This unwind information is needed for lldb gdb doesn't use it and tries
+    // to read all registers from $sp + 12
+    .save {r0-r15}
```

Comment:
Have you checked that the unwinder really restores the correct sp for the failure frame in LLDB? I am not 100% sure if the unwinder would use the SP register value it has restored from the stack or if it would ignore it and rather use the sp it computed by subtracting the size of the registers that were pushed here.
What I mean is to use a simple testing app that does a null reference, run it under lldb with breakpoint set to HandleHardwareException. When you hit the sigsegv, print the sp and then continue. When you hit the breakpoint in the HandleHardwareException, switch to the frame with the sigsegv and print the sp again. Verify that the two printed sp values match.

Reply:
Yes, I've checked it. Here lldb logs for testing app that use null reference:
[lldb log](https://gist.github.com/kbaladurin/c9c026837c48d2b96a3293a2f735d62d)
[lldb unwind log](https://gist.github.com/kbaladurin/62ce281d68a04056bf97fd0fc1dde8c6)

Frame with sigsegv is frame 3 and according to unwind log lldb restores `sp` register from the stack:
```
...
  th1/fr2 frame uses ARM.exidx unwind info for full UnwindPlan
  th1/fr2 active row: 0x00000000b2033cb0: CFA=sp+76 => r0=[CFA-64] r1=[CFA-60] r2=[CFA-56] r3=[CFA-52] r4=[CFA-48] r5=[CFA-44] r6=[CFA-40] r7=[CFA-36] r8=[CFA-32] r9=[CFA-28] r10=[CFA-24] r11=[CFA-20] r12=[CFA-16] sp=[CFA-12] lr=[CFA-8] pc=[CFA-4]
...
  th1/fr2 supplying caller's saved sp (13)'s location using ARM.exidx unwind info UnwindPlan
  th1/fr2 supplying caller's register sp (13) from the stack, saved at CFA plus offset -12 [saved at 0xbefff3d0]
   th1/fr3 sp = 0xbefff3f0
   th1/fr3 using architectural default unwind method
```

Reply:
Great, thank you!